### PR TITLE
Implement quirks for Samsung RIL in MSM822x et. al.

### PR DESCRIPTION
**Pull request by Flohack74**
This PR mimics necessary functional overrides that are normally done in Java code by Samsung to support their RIL interface extensions.
Specifically it tries to address:

- Carrier info
- Signal strength parsing
- Answering calls
- Making calls (with a specific patch for emergency calls that I cannot test for obvious reasons)

### CODE REVIEW

**Comment by UniversalSuperBox** [comment]
Is this still WIP?

**Comment by Flohack74** [comment]
Yes, sorry did not mark it.

**Comment by Flohack74** [comment]
Just found out you can create draft PRs, but you cannot switch a normal one back to draft. This is very frustrating...

**Comment by Flohack74** [comment]
So this is now working as far as I can tell on my Samsung device. I will give it a shot on my daily driver to see if there are any regressions for the normal RIL interface which I do not hope. Still testers would be very welcome!

**Comment by Flohack74** [comment]
Still ToDo:
- Use Android version in the driver name and make it more generic (the hope is it works for more than just MSM-822x devices)

File: gril/grilreply.c [hunk 1]
```diff
@@ -493,23 +493,35 @@ struct reply_data_reg_state *g_ril_reply_parse_data_reg_state(GRil *gril,
 
 	g_ril_append_print_buf(gril, "{");
 
+	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_8226) {
+		/* DANGER WILL ROBINSON
+         * In some cases from Vodaphone we are receiving a RAT of 102
+         * while in tunnels of the metro. Let's assume that if we
+         * receive 102 we actually want a RAT of 2 for EDGE service */
+/* 		if (str_arr->num_str > 4 && strcmp(str_arr->str[0], "1") && strcmp(str_arr->str[3], "102")) {
```
Comment (Flohack74) [inline gril/grilreply.c]: Not sure if we need that: It sounds more like a fix for Vodafone (and its not clear in which countries). I can include this if someone tells me how to replace the string in the struct - probably copying everything into a new one?

File: gril/grilreply.c [hunk 2]
```diff
@@ -493,23 +493,35 @@ struct reply_data_reg_state *g_ril_reply_parse_data_reg_state(GRil *gril,
...
+		case RST_IX_RAT:
 			set_reg_state(gril, &reply->reg_state, i, str);
 			break;
 		case RDST_IX_MAXDC:
 			set_data_reg_state(gril, reply, i, str);
 			break;
 		default:
 			g_ril_append_print_buf(gril, "%s%s", print_buf,
-						str ? str : "(null)");
+									str ? str : "(null)");
```
Comment (Flohack74) [inline gril/grilreply.c]: Will revert

**Comment by Flohack74** [comment]
So now this PR is ready for review ;)

File: drivers/rilmodem/network-registration.c [hunk 3]
```diff
@@ -238,7 +238,7 @@ static void ril_cops_cb(struct ril_msg *message, gpointer user_data)
 	struct ofono_network_operator op;
 
 	if (message->error != RIL_E_SUCCESS) {
-		ofono_error("%s: failed to retrive the current operator",
+		ofono_error("%s: failed to retrieve the current operator",
```
Comment (peat-psuwit) [inline drivers/rilmodem/network-registration.c]: Please leave this out to a separate PR.

File: gril/grilreply.c [hunk 4]
```diff
@@ -493,15 +493,29 @@ struct reply_data_reg_state *g_ril_reply_parse_data_reg_state(GRil *gril,
...
+         * In some cases from Vodaphone we are receiving a RAT of 102
+         * while in tunnels of the metro. Let's assume that if we
+         * receive 102 we actually want a RAT of 2 for EDGE service */
+/*	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x) {
+                if (str_arr->num_str > 4 && strcmp(str_arr->str[0], "1") && strcmp(str_arr->str[3], "102")) {
+			str_arr->str[3] = "2";
+		}
+	}
+*/
+
```
Comment (peat-psuwit) [inline gril/grilreply.c]: Is this still needed?
- If yes, I'm not sure if the assignment at line 503 works. I thought an `strcpy()` or manual `NULL` insertion is required.
- If no, please remove it.

File: gril/grilreply.c [hunk 5]
```diff
@@ -493,15 +493,29 @@ struct reply_data_reg_state *g_ril_reply_parse_data_reg_state(GRil *gril,
...
 		if (i > 0)
 			g_ril_append_print_buf(gril, "%s,", print_buf);
 
 		switch (i) {
-		case RST_IX_STATE: case RST_IX_LAC:
-		case RST_IX_CID:   case RST_IX_RAT:
+		case RST_IX_STATE:
+		case RST_IX_LAC:
+		case RST_IX_CID:
+		case RST_IX_RAT:
```
Comment (peat-psuwit) [inline gril/grilreply.c]: Not a functional change, please remove.

File: gril/grilreply.c [hunk 6]
```diff
@@ -852,7 +866,10 @@ GSList *g_ril_reply_parse_get_calls(GRil *gril, const struct ril_msg *message)
 	struct parcel rilp;
 	GSList *l = NULL;
 	int num, i;
-	gchar *number, *name;
+	int uusInfoPresent = 0;
+	int arrayLen;
+	gchar *number,
+		*name;
```
Comment (peat-psuwit) [inline gril/grilreply.c]: Please leave this line alone.

File: gril/grilreply.c [hunk 7]
```diff
@@ -894,8 +916,13 @@ GSList *g_ril_reply_parse_get_calls(GRil *gril, const struct ril_msg *message)
...
 
 		parcel_r_int32(&rilp); /* namePresentation */
-		parcel_r_int32(&rilp); /* uusInfo */
-
+		uusInfoPresent = parcel_r_int32(&rilp); /* uusInfo */
+		if (uusInfoPresent == 1) {
+			parcel_r_int32(&rilp); /* Type */
+			parcel_r_int32(&rilp); /* Dcs */
+			//Read byte array with user data and ignore it		
+			parcel_r_raw(&rilp, &arrayLen);
```
Comment (peat-psuwit) [inline gril/grilreply.c]: `parcel_r_raw()` returns a `g_try_malloc0()` pointer. Please don't forget to free it using `g_free()`.

File: gril/grilrequest.c [hunk 8]
```diff
@@ -807,24 +807,45 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
+          parcel_w_string(rilp, ""); // CallDetails.getCsvFromExtras
+      }
+      /* USS, empty string */
+      /* TODO: Deal with USS properly */
+      parcel_w_int32(rilp, 0);
+
+      g_ril_append_print_buf(gril, "(%s,%d,0,0)",
+                              phoneNoString,
+                              clir);
+}
```
Comment (peat-psuwit) [inline gril/grilrequest.c:835]: Seems like you inadvertently change the whitespace here. Remember, oFono uses tab characters, not spaces.

File: gril/grilrequest.h [hunk 9]
```diff
@@ -205,13 +205,17 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
 
 void g_ril_request_dial(GRil *gril,
-			const struct ofono_phone_number *ph,
-			enum ofono_clir_option clir,
-			struct parcel *rilp);
+                      const struct ofono_phone_number *ph,
+                      gboolean isEmergency,
+                      enum ofono_clir_option clir,
+                      struct parcel *rilp);
+
```
Comment (peat-psuwit) [inline gril/grilrequest.h:212]: Seems like you inadvertently change the whitespace here. Remember, oFono uses tab characters, not spaces.

File: gril/grilunsol.c [hunk 10]
```diff
@@ -533,12 +538,12 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
 
 	/* CDMA_SignalStrength */
 	cdma_dbm = parcel_r_int32(&rilp);
-	parcel_r_int32(&rilp); /* ecio */
+	parcel_r_int32(&rilp); /* cdmaEcio */
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: Non-functional change, please revert.

File: gril/ril_constants.h [hunk 11]
```diff
@@ -447,6 +447,7 @@
 #define RIL_UNSOL_DC_RT_INFO_CHANGED 1041
 #define RIL_UNSOL_RADIO_CAPABILITY 1042
 #define RIL_UNSOL_ON_SS 1043
+#define RIL_UNSOL_ON_SS_LL 11055
```
Comment (peat-psuwit) [inline gril/ril_constants.h]: If it's a vendor-specific constant, please put it in a vender-specific file. Sees https://github.com/ubports/ofono/blob/xenial/drivers/qcommsimmodem/qcom_msim_constants.h

File: include/voicecall.h [hunk 12]
```diff
@@ -140,9 +140,13 @@ void ofono_voicecall_en_list_notify(struct ofono_voicecall *vc,
...
 void ofono_voicecall_notify(struct ofono_voicecall *vc,
 				const struct ofono_call *call);
+
+int ofono_voicecall_is_emergency_number(struct ofono_voicecall *vc,
+				const struct ofono_phone_number *ph);
+
 void ofono_voicecall_disconnected(struct ofono_voicecall *vc, int id,
-				enum ofono_disconnect_reason reason,
-				const struct ofono_error *error);
+                              enum ofono_disconnect_reason reason,
```
Comment (peat-psuwit) [inline include/voicecall.h]: Seems like you inadvertently change whitespace here.

File: unit/test-grilrequest.c [hunk 13]
```diff
@@ -522,7 +522,7 @@ struct request_test_dial_data {
 static const guchar req_dial_parcel_valid_1[] = {
 	0x09, 0x00, 0x00, 0x00, 0x39, 0x00, 0x31, 0x00, 0x37, 0x00, 0x35, 0x00,
 	0x32, 0x00, 0x35, 0x00, 0x35, 0x00, 0x35, 0x00, 0x35, 0x00, 0x00, 0x00,
-	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
```
Comment (peat-psuwit) [inline unit/test-grilrequest.c]: Why the test case has to be edited here?

**Review by peat-psuwit — CHANGES_REQUESTED** [review]
Your code contains too many non-functional changes. They become noise in the review process.
- Unnecessary/incorrect whitespace changes.
- Unnecessary formatting changes.
- Unnecessary comment/string changes.

oFono follows the coding style of Linux. Thus, the first kind of change should be eliminated by running Linux's `checkpatch.pl` across your tree (which should also eliminate other formatting errors in your changes). For the other 2, please check my inline comments or checking the diff by yourself. Also, please read [oFono coding style document](https://github.com/ubports/ofono/blob/xenial/doc/coding-style.txt).

Some of my inline comments also contain some real functional changes. However, with too much noise on the change, I'm unable to look thoroughly.

File: gril/grilreply.c (hunk 4 above)
Comment (Flohack74) [reply gril/grilreply.c]: To be honest I have no idea if it is needed. Its in the original Java file like this. Line 503 probably does not work, but I will remove it entirely. Its also probably not a device-specific thing at all, they just found this and made a drive-by fix.

File: gril/ril_constants.h (hunk 11 above)
Comment (Flohack74) [reply gril/ril_constants.h]: The constant is unused ATM and I could also not find a real usage in Android/Lineage parts, so I removed it again.

**Comment by Flohack74** [comment]
I dont get it, my last commit is missing here. Please do not review now. Github is slow?

File: gril/grilreply.c [hunk 14]
```diff
@@ -853,6 +853,8 @@ GSList *g_ril_reply_parse_get_calls(GRil *gril, const struct ril_msg *message)
 	GSList *l = NULL;
 	int num, i;
 	gchar *number, *name;
+	int uusInfoPresent = 0;
+	int arrayLen;
```
Comment (peat-psuwit) [inline gril/grilreply.c]: oFono seems to prefer `snake_case` and [shortened](https://github.com/rilmodem/ofono/blob/efc9c0a85d32706bc088e449e847be41dcc73b3d/doc/coding-style.txt#L309) variable naming. Also, [don't initialize variable unnecessarily](https://github.com/rilmodem/ofono/blob/efc9c0a85d32706bc088e449e847be41dcc73b3d/doc/coding-style.txt#L147).

File: gril/grilrequest.c [hunk 15]
```diff
@@ -808,25 +808,47 @@ void g_ril_request_set_smsc_address(GRil *gril,
 
 void g_ril_request_dial(GRil *gril,
 			const struct ofono_phone_number *ph,
+			gboolean isEmergency,
 			enum ofono_clir_option clir,
 			struct parcel *rilp)
 {
+	const char *phoneNoString;
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: Similar codes name this variable simply `number`. Also, maybe moving initialization to the same line as declaration?

File: gril/grilrequest.c [hunk 16]
```diff
@@ -808,25 +808,47 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
 	parcel_w_int32(rilp, clir);
+	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x) {
+		parcel_w_int32(rilp, 0); // CallDetails.call_type
+		if (isEmergency) {
+			parcel_w_int32(rilp, 3); // Emergency call domain?
+		} else {
+			parcel_w_int32(rilp, 1); // Default call domain?
+		}
+		parcel_w_string(rilp, ""); // CallDetails.getCsvFromExtras
+	}
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: I would do something like:
```c
int call_domain = is_emergency ? 3 : 1;
parcel_w_int32(rilp, 0); // CallDetails.call_type
parcel_w_int32(rilp, call_domain); // CallDetails.call_domain
parcel_w_string(rilp, ""); // CallDetails.getCsvFromExtras
```

File: gril/grilrequest.c [hunk 17]
```diff
@@ -808,25 +808,47 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
 	parcel_w_int32(rilp, 0);
 
 	g_ril_append_print_buf(gril, "(%s,%d,0,0)",
-				phone_number_to_string(ph),
+				phoneNoString,
 				clir);
 }
 
+void g_ril_request_answer(GRil *gril,
+						struct parcel *rilp)
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: You don't have to indent until hitting 80 characters. Just indent it past the opened parenthesis should be enough (see `g_ril_request_hangup` below).

File: gril/grilunsol.c [hunk 18]
```diff
@@ -519,11 +519,16 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
 	int lte_signal;
 	int signal;
 
+	gboolean samsungQuirks = g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x;
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: `snake_case`. Also, wrapping it in parentheses should improve readability.

File: gril/grilunsol.c [hunk 19]
```diff
@@ -519,11 +519,16 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
...
+
 	g_ril_init_parcel(message, &rilp);
 
 	/* RIL_SignalStrength_v5 */
 	/* GW_SignalStrength */
 	gw_sigstr = parcel_r_int32(&rilp);
+	if (samsungQuirks) {
+		gw_sigstr &= 0xFF;
+	}
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: [Do not unnecessarily use braces where a single statement will do](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces). Also, I see that there's no comment about why this is done in the RIL.java, but a comment stating so should reduce confusion.

What happens if this is not done?

File: gril/grilunsol.c [hunk 20]
```diff
@@ -548,7 +553,17 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
...
 		parcel_r_int32(&rilp); /* cqi */
+		if (samsungQuirks) {
+			if ((lte_sigstr & 0xff) == 255 || lte_sigstr == 99) {
+				lte_sigstr = 99;
+				lte_rsrp = INT_MAX;
+				lte_rssnr = INT_MAX;
+			} else {
+				lte_sigstr &= 0xff;
+			}
+		}
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: My comment about the comment applies here too.

File: include/voicecall.h [hunk 21]
```diff
@@ -140,6 +140,10 @@ void ofono_voicecall_en_list_notify(struct ofono_voicecall *vc,
 
 void ofono_voicecall_notify(struct ofono_voicecall *vc,
 				const struct ofono_call *call);
+
+int ofono_voicecall_is_emergency_number(struct ofono_voicecall *vc,
+				const struct ofono_phone_number *ph);
```
Comment (peat-psuwit) [inline include/voicecall.h]: I'm not sure about the naming & the arguments of this function. On naming, I guess adding 2 underscores in front of it might be a good idea. As for arguments, another possibility is to pass the string in (which is essentially exposing the original `is_emergency_number` in `voicecall.c`). @UniversalSuperBox, could you please comment?

(Also, what if I invited the original engineer at Canonical to have a look at this?)

File: plugins/samsung_msm_822x.c [hunk 22]
```diff
@@ -0,0 +1,78 @@
+#ifdef HAVE_CONFIG_H
```
Comment (peat-psuwit) [inline plugins/samsung_msm_822x.c:1]: This file is still using spaces instead of tabs.

File: plugins/samsung_msm_822x.c [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+    }
+}
+
+static const char *samsung_msm_822x_request_id_to_string(int req)
+{
+    switch (req) {
+    default:
+        return NULL;
+    }
+}
```
Comment (peat-psuwit) [inline plugins/samsung_msm_822x.c]: These functions aren't needed. You can pass `NULL` in place of them in `ril_create`.

File: unit/test-grilrequest.c [hunk 24]
```diff
@@ -1213,7 +1213,7 @@ static void test_request_dial(gconstpointer data)
 	enum ofono_clir_option clir = test_data->clir;
 	struct parcel rilp;
 
-	g_ril_request_dial(NULL, ph, clir, &rilp);
+	g_ril_request_dial(NULL, ph, FALSE, clir, &rilp);
```
Comment (peat-psuwit) [inline unit/test-grilrequest.c:1216]: Hmm, maybe a testcase for your gril modification here?

File: drivers/rilmodem/voicecall.c [hunk 25]
```diff
@@ -598,10 +599,19 @@ static void ril_ss_notify(struct ril_msg *message, gpointer user_data)
...
 
+	g_ril_request_answer(vd->ril, &rilp);
+
 	/* Send request to RIL */
-	ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, NULL, cb, data);
+	if (rilp.size > 0) {
+		ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, &rilp, cb, data);
+	} else {
+		ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, NULL, cb, data);
+	}
```
Comment (peat-psuwit) [inline drivers/rilmodem/voicecall.c]: [Do not unnecessarily use braces where a single statement will do](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces).

**Review by peat-psuwit — CHANGES_REQUESTED** [review]
Most of the reviews this time are still formatting changes. However, the changes are cleaner and can be reviewed more thoroughly. Still, running `checkpatch.pl` across all files (and let it fixes fixable problems) is always a good idea.

On the technical side, I don't have much problem with it. However, I'm not sure about exposing is_emergency_number() outside of `voicecall.c`. The ultimate goal is (or at least was) to merge this into upstream oFono. So, I'm not sure if oFono maintainers will be ok with it. Who should I ask/contact in this case?

File: gril/grilrequest.c (hunk 17 above)
Comment (Flohack74) [reply gril/grilrequest.c]: I did not do this on purpose ;)

File: include/voicecall.h (hunk 21 above)
Comment (Flohack74) [reply include/voicecall.h]: You can ask anyone you like for additional review, more eyes the better...

File: drivers/rilmodem/voicecall.c (hunk 25 above)
Comment (Flohack74) [reply drivers/rilmodem/voicecall.c]: I am sorry, this comes from a mandatory rule in my company, I have to do this all day ;)

File: unit/test-grilrequest.c (hunk 24 above)
Comment (Flohack74) [reply unit/test-grilrequest.c:1216]: The problem is that the codepath where the emergency flag is being evaluated only activates when the plugin is detected. All current test cases there have `*gril` set to NULL, meaning it will never find a plugin type and also not switch to Samsung mode. To make better test cases this would involve mocking the result of the `g_ril_vendor(gril)` call which I am not able to do, honestly.

File: include/voicecall.h (hunk 21 above)
Comment (Flohack74) [reply include/voicecall.h]: I decided to expose the existing function to not make unnecessary nesting. See update

File: gril/grilreply.c [hunk 26]
```diff
@@ -894,8 +901,13 @@ GSList *g_ril_reply_parse_get_calls(GRil *gril, const struct ril_msg *message)
 		}
 
 		parcel_r_int32(&rilp); /* namePresentation */
-		parcel_r_int32(&rilp); /* uusInfo */
-
+		has_uusinfo = parcel_r_int32(&rilp); /* uusInfo */
+		if (has_uusinfo == 1) {
+			parcel_r_int32(&rilp); /* Type */
+			parcel_r_int32(&rilp); /* Dcs */
+			//Read byte array with user data and ignore it
```
Comment (peat-psuwit) [inline gril/grilreply.c]: Remove trailing whitespace.

```suggestion
			//Read byte array with user data and ignore it
```

File: drivers/rilmodem/voicecall.c [hunk 27]
```diff
@@ -598,10 +599,18 @@ static void ril_ss_notify(struct ril_msg *message, gpointer user_data)
...
 	DBG("Answering current call");
 
+	g_ril_request_answer(vd->ril, &rilp);
+
 	/* Send request to RIL */
-	ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, NULL, cb, data);
+	if (rilp.size > 0)
+		ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, &rilp, cb, data);
+	else
+		ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0, NULL, cb, data);
```
Comment (peat-psuwit) [inline drivers/rilmodem/voicecall.c]: I guess this can be re-written to:
```suggestion
	ril_template(RIL_REQUEST_ANSWER, vc, generic_cb, 0,
				rilp.size > 0 ? &rilp : NULL, cb, data);
```

File: gril/grilrequest.c [hunk 28]
```diff
@@ -808,25 +808,43 @@ void g_ril_request_set_smsc_address(GRil *gril,
 
 void g_ril_request_dial(GRil *gril,
 			const struct ofono_phone_number *ph,
+			gboolean is_emergency,
 			enum ofono_clir_option clir,
 			struct parcel *rilp)
 {
+	const char *number = phone_number_to_string(ph);
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: I think this refactor is not needed?

File: gril/grilunsol.c [hunk 29]
```diff
@@ -519,11 +519,16 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
...
 	int signal;
 
+	gboolean samsung_quirks = (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x);
+
 	g_ril_init_parcel(message, &rilp);
 
 	/* RIL_SignalStrength_v5 */
 	/* GW_SignalStrength */
 	gw_sigstr = parcel_r_int32(&rilp);
+	/* Samsung on MSM822x does need a weird masking here. From vendor's java file. */
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: "Vendor's java file" implies that you have access to Samsung's source code (which you don't). Refers to LOS's code instead (and preferably the link).

```suggestion
	/*
	 * Samsung on MSM822x needs weird masking here, according to LOS's RIL
	 * for this device.
	 * https://github.com/LineageOS/android_device_samsung_msm8226-common/blob/cm-12.1/ril/SamsungMSM8226RIL.java
	 */
```

File: gril/grilunsol.c [hunk 30]
```diff
@@ -548,7 +553,18 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
 		parcel_r_int32(&rilp); /* rsrq */
 		lte_rssnr = parcel_r_int32(&rilp);
 		parcel_r_int32(&rilp); /* cqi */
+		/* Samsung on MSM822x calculation of signal strength. From vendor's java file. */
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: Same as the above (but I guess URL can be left this time :) )

File: drivers/rilmodem/voicecall.c [hunk 31]
```diff
@@ -414,7 +414,8 @@ static void dial(struct ofono_voicecall *vc,
 	struct cb_data *cbd = cb_data_new(cb, data, vc);
 	struct parcel rilp;
 
-	g_ril_request_dial(vd->ril, ph, clir, &rilp);
+	//Samsung wants emergency number indicated also in the parcel to rild
+	g_ril_request_dial(vd->ril, ph, ofono_voicecall_is_emergency_number(vc, phone_number_to_string(ph)), clir, &rilp);
```
Comment (peat-psuwit) [inline drivers/rilmodem/voicecall.c]: Line over 80 characters.

File: gril/grilrequest.c [hunk 32]
```diff
@@ -808,25 +808,43 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
 	parcel_init(rilp);
 
 	/* Number to dial */
-	parcel_w_string(rilp, phone_number_to_string(ph));
+	parcel_w_string(rilp, number);
 	/* CLIR mode */
 	parcel_w_int32(rilp, clir);
+	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x) {
+		parcel_w_int32(rilp, 0); // CallDetails.call_type
+		parcel_w_int32(rilp, call_domain); // Normal or emergency call domain
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: Line over 80 characters.

Maybe this?

```suggestion
		parcel_w_int32(rilp, call_domain); // Normal or emergency
```

File: gril/grilunsol.c [hunk 33]
```diff
@@ -519,11 +519,16 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
 	int lte_signal;
 	int signal;
 
+	gboolean samsung_quirks = (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x);
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: Line over 80 characters.

File: plugins/samsung_msm_822x.c [hunk 34]
```diff
@@ -0,0 +1,68 @@
...
+		.enable = ril_enable,
+		.disable = ril_disable,
+		.pre_sim = ril_pre_sim,
+		.post_sim = ril_post_sim,
+		.post_online = ril_post_online,
+		.set_online = ril_set_online,
+};
+
+/*
+ * This plugin is a device plugin for Samsung's devices with MSM-822x baseband that use
```
Comment (peat-psuwit) [inline plugins/samsung_msm_822x.c]: Line over 80 characters.

**Review by peat-psuwit — CHANGES_REQUESTED** [review]
Running `checkpatch.pl` locally, there are still some style issues. In particular, there's trailing whitespace in `grilreply.c` and there are lines that go over 80 characters.

Also, a few other changes.

File: gril/grilrequest.c (hunk 32 above)
Comment (Flohack74) [reply gril/grilrequest.c]: This is exactly why I do not like tabs to indent - it will cause confusion with the line length and columns: On my IDE this line below 80 characters because my tab size is 4 spaces. How should I guess this from the code base alone (No knowledge about the original settings)? I can now just _assume_ your tab size might be 8 characters? Or 6? I don't know. Tabs are evil these days.

Comment (peat-psuwit) [reply gril/grilrequest.c]: [Tabs are 8 characters, and thus indentations are also 8 characters.](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#indentation).

Comment (Flohack74) [reply gril/grilrequest.c]: Well I am not going to discuss this further here, and this project is not a part of the Linux kernel, so any rule might or might not apply. I just say I simply cannot know it, so I need to guess. In my company all tabs are 4 characters for example, it is in my DNA so to say. So when I started to fix the overly long lines, I had to do it now 2 times. Any how, its fixed now...

Comment (peat-psuwit) [reply gril/grilrequest.c]: [oFono coding style must follow every rule for Linux kernel](https://github.com/ubports/ofono/blob/xenial/doc/coding-style.txt)

File: gril/grilunsol.c [hunk 35]
```diff
@@ -519,11 +520,20 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
...
+
 	g_ril_init_parcel(message, &rilp);
 
 	/* RIL_SignalStrength_v5 */
 	/* GW_SignalStrength */
 	gw_sigstr = parcel_r_int32(&rilp);
+	// Samsung on MSM822x does need a weird masking here, according to
+	// LineageOS's RIL adaptor taken from https://github.com/LineageOS/
+	// android_device_samsung_msm8226-common/blob/cm-12.1/ril/
+	// SamsungMSM8226RIL.java
```
Comment (peat-psuwit) [inline gril/grilunsol.c]:
```c
/*
 * Comment multi-line like this. See
 * https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting
 *
 * I did gave a recommendation for this, right? Sorry for forgetting about this
 * rule.
 */
```

File: gril/grilunsol.c [hunk 36]
```diff
@@ -548,7 +558,19 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
 		parcel_r_int32(&rilp); /* rsrq */
 		lte_rssnr = parcel_r_int32(&rilp);
 		parcel_r_int32(&rilp); /* cqi */
+		/* Samsung on MSM822x calculation of signal strength,
+		according to LineageOS's RIL adaptor */
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: Here, too.

File: gril/grilunsol.c (hunk 35 above)
Comment (peat-psuwit) [reply gril/grilunsol.c]: And I'm sorry this comment has been in the pending state for months.

**Review by peat-psuwit — CHANGES_REQUESTED** [review]
I'm really sorry, but I forgot to submit a pending comment about formatting. :frowning: 

Meanwhile, I'll take a look at the logic in more detail.

File: gril/grilunsol.c (hunk 35 above)
Comment (peat-psuwit) [reply gril/grilunsol.c]: Oh, BTW, please don't split the link across lines. If it's long, put it in its own line.

File: gril/grilrequest.c [hunk 37]
```diff
@@ -808,15 +808,22 @@ void g_ril_request_set_smsc_address(GRil *gril,
 
 void g_ril_request_dial(GRil *gril,
 			const struct ofono_phone_number *ph,
+			gboolean is_emergency,
 			enum ofono_clir_option clir,
 			struct parcel *rilp)
 {
+	int call_domain = is_emergency ? 3 : 1;
```
Comment (peat-psuwit) [inline gril/grilrequest.c]: I think this should be moved below so that it's clearer that it's Samsung-specific.

```suggestion
```

File: gril/grilrequest.c [hunk 38]
```diff
@@ -808,15 +808,22 @@ void g_ril_request_set_smsc_address(GRil *gril,
...
 
 	/* Number to dial */
 	parcel_w_string(rilp, phone_number_to_string(ph));
 	/* CLIR mode */
 	parcel_w_int32(rilp, clir);
+	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x) {
+		parcel_w_int32(rilp, 0); // CallDetails.call_type
+		parcel_w_int32(rilp, call_domain); // Normal or emergency
+		parcel_w_string(rilp, ""); // CallDetails.getCsvFromExtras
+	}
```
Comment (peat-psuwit) [inline gril/grilrequest.c]:
```suggestion
	if (g_ril_vendor(gril) == OFONO_RIL_VENDOR_SAMSUNG_MSM_822x) {
		int call_domain = is_emergency ? 3 : 1;
		parcel_w_int32(rilp, 0); // CallDetails.call_type
		parcel_w_int32(rilp, call_domain); // CallDetails.call_domain
		parcel_w_string(rilp, ""); // CallDetails.getCsvFromExtras
	}
```

File: gril/grilunsol.c [hunk 39]
```diff
@@ -537,7 +537,8 @@ static int get_gsm_strength(int signal)
 		return -1;
 }
 
-int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
+int g_ril_unsol_parse_signal_strength(GRil *gril,
+					const struct ril_msg *message,
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: I don't think this needs to be changed? The line was _exactly_ 80-chars long.

File: gril/grilunsol.c [hunk 40]
```diff
@@ -575,7 +585,19 @@ int g_ril_unsol_parse_signal_strength(GRil *gril, const struct ril_msg *message,
...
+			if ((lte_sigstr & 0xff) == 255 || lte_sigstr == 99) {
+				lte_sigstr = 99;
+				lte_rsrp = INT_MAX;
+				lte_rssnr = INT_MAX;
+			} else {
+				lte_sigstr &= 0xff;
+			}
+		}
 		lte_signal = get_lte_strength(lte_sigstr, lte_rsrp, lte_rssnr);
+		parcel_r_int32(&rilp); /* tdScdmaRscp */
```
Comment (peat-psuwit) [inline gril/grilunsol.c]: Is this specific to Samsung or just a value that has never been read?

**Review by peat-psuwit — CHANGES_REQUESTED** [review]

File: gril/grilunsol.c (hunk 40 above)
Comment (Flohack74) [reply gril/grilunsol.c]: This was never read and I suggest to read it off the parcel, just for completeness.

Comment (peat-psuwit) [reply gril/grilunsol.c]: This field seems to be added since RIL version 10.
https://github.com/LineageOS/android_hardware_ril/blob/cm-14.1/include/telephony/ril.h#L1257

You'll have to properly guard this read under either `parcel_data_avail()` or RIL version check. In fact, in RIL version 8 there's a field between this and `tdScdmaRscp`: `lteTimingAdvance`. As this isn't required for the Samsung MSM822x devices, and this isn't a part of an "array", sets of the same data type repeating a number of times, I would say this should be removed.

**Review by peat-psuwit — CHANGES_REQUESTED** [review]

File: gril/grilunsol.c (hunk 40 above)
Comment (Flohack74) [reply gril/grilunsol.c]: Removed!

**Review by peat-psuwit — APPROVED** [review]
I think this is now OK to merge. I've tested calling on E4.5 and Sony Xperia X and does not found a regression.

However, given how fundamental this component is and the OTA-14 freeze coming soon, I would suggest merging this after the OTA-14 cycle.
